### PR TITLE
os/bluestore: BufferSpace per Onode

### PR DESCRIPTION
**Pull request by pereman2**
Instead of indexing buffers by blob offset we do it by logical offset. Previously we tried using physical offsets but it was a pretty bad idea is it required extra bound checks of physical extents and extra loops for to add them, that includes chopping off buffer data.

Logical offsets don't have these problems but we have different ones. Bufferlist data is padded before being used in `_do_alloc_write` which means before adding it to BufferSpace, we are required to get the unpadded substr_of data because padded data might overflow to an invalid logical offset in contrast to when we used blob offsets where padded data was wonderful. This is true in in `generate_read_result_bl` too, we read by regions of blob extents but we must translate it back to logical offsets, that means an extra loop to include correct logical regions was added.

BufferSpace::read now has an extra early return check that makes sure that there is an overlap between what we want to read and the buffers we get.

This pr also includes some identation fixes because it's driving me nuts.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by pereman2** [comment]
We ran various test and IMHO there is no clear improvement a part from some minor IOPS improvement in `randread` workloads. 

These are the results:
**FIO of combined randread, randrw and randwrite workloads**
image

**Buffer perf counters**
image

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -295,17 +298,21 @@ class BlueStore : public ObjectStore,
     uint16_t cache_private = 0; ///< opaque (to us) value used by Cache impl
     uint32_t flags;             ///< FLAG_*
     uint64_t seq;
-    uint32_t offset, length;
+    uint64_t offset, length;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.h:301]: There is no need to go 64 bits.
Max object size in BlueStore is 32 bits.
I interpret that in the way, that the offset of the last accessible byte of object is < 2^32.

https://github.com/ceph/ceph/blob/main/src/os/bluestore/BlueStore.cc#L145

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -374,8 +381,8 @@ class BlueStore : public ObjectStore,
     state_list_t writing;   ///< writing buffers, sorted by seq, ascending
 
     ~BufferSpace() {
-      ceph_assert(buffer_map.empty());
       ceph_assert(writing.empty());
+      ceph_assert(buffer_map.empty());
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.h:385]: Line swap unneeded.

File: src/os/bluestore/BlueStore.h [hunk 3]
```diff
@@ -407,11 +414,15 @@ class BlueStore : public ObjectStore,
...
+    std::map<uint64_t, std::unique_ptr<Buffer>>::iterator
     _rm_buffer(BufferCacheShard* cache,
-		    std::map<uint32_t, std::unique_ptr<Buffer>>::iterator p) {
+		    std::map<uint64_t, std::unique_ptr<Buffer>>::iterator p) {
       ceph_assert(p != buffer_map.end());
+      std::stringstream ss;
+      ss << __func__ << std::hex << " 0x" << p->first << "~"
+                            << p->second->length << std::dec << std::endl;
       cache->_audit("_rm_buffer start");
+      cache->_audit(ss.str().c_str());
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.h:425]: Doing audit 2 times in a row. 
Having custom name for audit reason is nice, but then wrap it with #ifdef DEBUG_CACHE.

File: src/os/bluestore/BlueStore.h [hunk 4]
```diff
@@ -454,25 +465,25 @@ class BlueStore : public ObjectStore,
...
       cache->_trim();
     }
-    void _finish_write(BufferCacheShard* cache, uint64_t seq);
-    void did_read(BufferCacheShard* cache, uint32_t offset, ceph::buffer::list& bl) {
+    void _finish_write(BufferCacheShard* cache, uint64_t offset, uint64_t seq);
+    void did_read(BufferCacheShard* cache, uint64_t offset, uint64_t length, ceph::buffer::list&& bl) {
       std::lock_guard l(cache->lock);
-      Buffer *b = new Buffer(this, Buffer::STATE_CLEAN, 0, offset, bl);
-      b->cache_private = _discard(cache, offset, bl.length());
+      Buffer *b = new Buffer(this, Buffer::STATE_CLEAN, 0, offset, std::move(bl));
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.h:480]: Interesting. Has this improvement been possible before?

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -1851,8 +1846,13 @@ void BlueStore::BufferSpace::read(
   cache->logger->inc(l_bluestore_buffer_miss_bytes, miss_bytes);
 }
 
-void BlueStore::BufferSpace::_finish_write(BufferCacheShard* cache, uint64_t seq)
+void BlueStore::BufferSpace::_finish_write(BufferCacheShard* cache, uint64_t offset, uint64_t seq)
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:1931]: `offset` looks like a debug addition, to remove?

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -1693,24 +1663,33 @@ BlueStore::BufferCacheShard *BlueStore::BufferCacheShard::create(
 #undef dout_prefix
 #define dout_prefix *_dout << "bluestore.BufferSpace(" << this << " in " << cache << ") "
 
-void BlueStore::BufferSpace::_clear(BufferCacheShard* cache)
+bool BlueStore::BufferSpace::_dup_writing(TransContext* txc, BufferCacheShard* cache, OnodeRef onode)
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc]: Was changing a location (and dropping func description) of _dup_writing deliberate?
If not, its better to minimize diff.

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -1695,6 +1695,12 @@ int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint64_t offset, u
     if (b->offset >= end) {
       break;
     }
+    // If no overlap is found between what we want to read and what we have
+    // cached, then forget about continuing
+    if (std::max(b->end(), offset + length) - std::min(b->offset, offset) >
+        length + b->length) {
+      break;
+    }
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:1779]: `i` from `_data_lower_bound()` either gives buffer that contains offset, or first one after.
I do not understand what this condition is doing.
Did you have in mind a case that `b->offset < end` but also `b->offset + b->length < offset` ?
That would be assert against `_data_lower_bound()` guarantee.

File: src/os/bluestore/BlueStore.h (hunk 4 above)
Comment (pereman2) [reply src/os/bluestore/BlueStore.h:475]: I don't think so. This buffer is being moved because we have to substr the previously padded data, therefore this case can just take ownership of the bl passed.
The other case has to copy the padded buffer so that it's stored in buffer.

I believe we could make improvements by following a close to 0 copy of bufferlist data by just reusing buffers created while doing a write.

For example in the context of main branch: we do `_do_alloc_write` and we perform two sequential operations like this:

```c++
    _buffer_cache_write(txc, wi.b, o, wi.logical_offset, wi.bl,
                        wctx->buffered ? 0 : Buffer::FLAG_NOCACHE);
    wi.b->get_blob().map_bl(
	  b_off, *l,
	  [&](uint64_t offset, bufferlist& t) {
          bdev->aio_write(offset, t, &txc->ioc, false);
   });
``` 

If you think about it here we have 1 copy of `wi.bl` inside `_buffer_cache_write` when doing `new Buffer`. It is completely unnecessary, we could just do something like:

```
    Buffer* buffer =_buffer_cache_write(txc, wi.b, o, wi.logical_offset, wi.bl,
                        wctx->buffered ? 0 : Buffer::FLAG_NOCACHE);
    wi.b->get_blob().map_bl(
	  b_off, *buffer->data, // <<<<<<<<<<<<<<<<<< reference of data
	  [&](uint64_t offset, bufferlist& t) {
          bdev->aio_write(offset, t, &txc->ioc, false);
   });
```

File: src/os/bluestore/BlueStore.cc (hunk 5 above)
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:1936]: yes, will be removed, it is a spillover of previous failed implementations

File: src/os/bluestore/BlueStore.cc (hunk 6 above)
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc]: I will move that back

**Comment by pereman2** [comment]
https://pulpito.ceph.com/pdiazbou-2024-01-29_15:40:07-rados-wip-pereman2-buffer-space-distro-default-smithi/

**Comment by aclamk** [comment]
jenkins test windows tests

**Comment by aclamk** [comment]
jenkins test windows

**Comment by aclamk** [comment]
jenkins test api

**Comment by aclamk** [comment]
jenkins test api

**Comment by aclamk** [comment]
jenkins test windows

**Comment by aclamk** [comment]
jenkins test api

File: src/os/bluestore/BlueStore.cc [hunk 8]
```diff
@@ -1584,7 +1585,7 @@ struct TwoQBufferCacheShard : public BlueStore::BufferCacheShard {
 #ifdef DEBUG_CACHE
   void _audit(const char *when) override
   {
-    dout(10) << __func__ << " " << when << " start" << dendl;
+    dout(10) << __func__ << " start" << dendl;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:1588]: please add `when` back

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -1660,13 +1661,14 @@ void BlueStore::BufferSpace::_clear(BufferCacheShard* cache)
...
   }
+  writing.clear();
 }
 
 int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint32_t offset, uint32_t length)
 {
   // note: we already hold cache->lock
-  ldout(cache->cct, 20) << __func__ << std::hex << " 0x" << offset << "~" << length
-           << std::dec << dendl;
+  ldout(cache->cct, 20) << __func__ << std::hex << " 0x" << offset << "~"
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:1670]: Please undo this unneeded change.

File: src/os/bluestore/BlueStore.cc (hunk 7 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:1686]: This disjointment test seems unnecessary.

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -1750,7 +1758,9 @@ void BlueStore::BufferSpace::read(
     for (auto i = _data_lower_bound(offset);
          i != buffer_map.end() && offset < end && i->first < end; ++i) {
       Buffer *b = &i->second;
-      ceph_assert(b->end() > offset);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:1753]: The assert should remain here.

We assume no buffers overlap, and buffers are processed in order.
On first iteration, with i from `_data_lower_bound()` condition holds:
`offset < b->end()` but we cannot know whether `offset <= b->offset` or `offset > b->offset`
On each subsequent iteration conditions holds:
`offset <= b->offset` and `offset < b->end()`

So the assert checks our assumptions on buffers.

File: src/os/bluestore/BlueStore.cc [hunk 11]
```diff
@@ -3122,44 +3109,42 @@ void BlueStore::ExtentMap::make_range_shared_maybe_merge(
...
-	       << std::dec << dendl;
-      Blob* b = blob.is_compressed() ? nullptr :
-	find_mergable_companion(e.blob.get(), e.blob_start(), blob_width, candidates);
+      uint32_t blob_width; // to signal when extents end
+      dout(20) << __func__ << std::hex << " e.blob_start=" << e.blob_start()
+               << " e.logical_offset=" << e.logical_offset << std::dec << dendl;
+      Blob *b = blob.is_compressed()
+                    ? nullptr
+                    : find_mergable_companion(e.blob.get(), e.blob_start(),
+                                              blob_width, candidates);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:3125]: dout lines are improved, but splitting `Blob* b` assignment to 4 lines looks bad

File: src/os/bluestore/BlueStore.cc [hunk 12]
```diff
@@ -3122,44 +3109,42 @@ void BlueStore::ExtentMap::make_range_shared_maybe_merge(
...
-	  if (p.is_valid()) {
-	    b->get_shared_blob()->get_ref(p.offset, p.length);
-	  }
-	}
-	// reblob extents might erase e
-	dirty_range_end = std::max<uint32_t>(dirty_range_end, e.blob_start() + b_logical_length);
-	uint32_t goto_logical_offset = e.logical_offset + e.length;
-	reblob_extents(e.blob_start(), e.blob_start() + blob_width,
+        dout(20) << __func__ << " merging to: " << *b << " bc=" << onode->bc
+                 << dendl;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:3128]: meld `<< dendl` to prev line.

File: src/os/bluestore/BlueStore.cc [hunk 13]
```diff
@@ -3794,40 +3771,40 @@ void BlueStore::ExtentMap::reshard(
...
-	  if (must_span) {
+        if (!e->blob->is_spanning()) {
+          // We have two options: (1) split the blob into pieces at the
+          // shard boundaries (and adjust extents accordingly), or (2)
+          // mark it spanning.  We prefer to cut the blob if we can.  Note that
+          // we may have to split it multiple times--potentially at every
+          // shard boundary.
+          bool must_span = false;
+          BlobRef b = e->blob;
+          if (onode->bc.writing.empty() && b->can_split()) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:3782]: Why is having `writing` buffers active preventing from splitting a blob ?

File: src/os/bluestore/BlueStore.cc [hunk 14]
```diff
@@ -5182,10 +5149,23 @@ void BlueStore::Collection::split_cache(
...
       }
       for (auto& b : o->extent_map.spanning_blob_map) {
-	b.second->last_encoded_id = -1;
+        b.second->last_encoded_id = -1;
+      }
+      // By default do not copy buffers to clones, and let them read data by
+      // themselves. The exception are 'writing' buffers, which are not yet
+      // stable on device.
+      for (auto &i : o->bc.buffer_map) {
+        if (!i.second.is_writing()) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5161]: Why cannot we also copy `writing` buffers? I was under impresion that new `finish_write` targets Onode, not blobs.

File: src/os/bluestore/BlueStore.cc [hunk 15]
```diff
@@ -2857,49 +2881,13 @@ uint32_t BlueStore::Blob::merge_blob(CephContext* cct, Blob* blob_to_dissolve)
...
-  // move BufferSpace writing
-  auto wrt_dst_it = dst->bc.writing.begin();
-  while(!src->bc.writing.empty()) {
-    Buffer& buf = src->bc.writing.front();
-    src->bc.writing.pop_front();
-    while (wrt_dst_it != dst->bc.writing.end() && wrt_dst_it->seq < buf.seq) {
-      ++wrt_dst_it;
-    }
-    dst->bc.writing.insert(wrt_dst_it, buf);
-  }
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:2877]: Onodes are the overlord of buffers now, buffers don't need to be moved as on operations on blobs.

[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:2877]: But those we writing buffers. If we drop them, we do not have them.

[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:2877]: We are not dropping them, we still have them in the same place. Blob src and Blob dst are in the same onode

File: src/os/bluestore/BlueStore.cc [hunk 16]
```diff
@@ -16327,7 +16347,7 @@ void BlueStore::_wctx_finish(
     // longer allocated.  Note that this will leave behind edge bits
     // that are no longer referenced but not deallocated (until they
     // age out of the cache naturally).
-    b->discard_unallocated(c.get());
+    // b->discard_unallocated(c.get(), o.get(), lo.e.logical_offset);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc]: Thanks for pointing it out, this was a memory leak that surprisingly didn’t affect the logic because writes just overwrite these buffers. discard_unallocated is enefficient with newest semantics because it would have to find overlaping segments between and old extent and pexents. I changed it so that truncate and zero do really remove buffers related to removed range since other punch_hole cases like write do just refill punched extent

File: src/os/bluestore/BlueStore.cc (hunk 7 above)
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:1684]: I've modified it to work with overlapping functions (the previous formula was terrible idk how I got past kindergarten math courses), imho it's easier to think about overlapping ranges.

File: src/os/bluestore/BlueStore.h [hunk 17]
```diff
@@ -689,8 +704,7 @@ class BlueStore : public ObjectStore,
     bool can_split() {
       std::lock_guard l(get_cache()->lock);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.h]: yes!

File: src/os/bluestore/BlueStore.cc [hunk 18]
```diff
@@ -13648,10 +13649,21 @@ void BlueStore::_txc_finish(TransContext *txc)
   dout(20) << __func__ << " " << txc << " onodes " << txc->onodes << dendl;
   ceph_assert(txc->get_state() == TransContext::STATE_FINISHING);
 
-  for (auto& sb : txc->blobs_written) {
-    sb->finish_write(txc->seq);
+  for (auto &[onode, seq] : txc->buffers_written) {
+    if (txc->deferred_txn && txc->deferred_txn->txc_seq == seq) {
+      std::unique_lock l(onode->c->lock);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc]: added the protection back

File: src/os/bluestore/BlueStore.h [hunk 19]
```diff
@@ -1638,6 +1641,11 @@ class BlueStore : public ObjectStore,
     pool_opts_t pool_opts;
     ContextQueue *commit_queue;
 
+    // Deferred write dependencies. Buffers of deferred writes than get cloned must be marked as clean
+    // on all cloned Onodes after the main deferred txc is completed.
+    std::mutex deferred_seq_dependencies_lock;
+    std::map<uint64_t, std::set<OnodeRef>> deferred_seq_dependencies;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:1647]: What happens with this collection when we split collection/cache. Doesn't it need splitting too?

Comment (pereman2) [reply src/os/bluestore/BlueStore.h:1647]: Buffers are addressed by onode and the onode is move to its respective collection so I think this is ok.

File: src/os/bluestore/BlueStore.cc [hunk 20]
```diff
@@ -5182,10 +5115,16 @@ void BlueStore::Collection::split_cache(
...
       }
       for (auto& b : o->extent_map.spanning_blob_map) {
-	b.second->last_encoded_id = -1;
+        b.second->last_encoded_id = -1;
+      }
+
+      for (auto &i : o->bc.buffer_map) {
+        ldout(store->cct, 1)
+          << __func__ << "   moving " << i.second << dendl;
+        dest->cache->_move(cache, &i.second);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:5127]: You still need those buffers in the BufferSpace to keep referring to them, so I think it made sense to move them all? I don't know if I missed some edge cases here

File: src/os/bluestore/BlueStore.cc [hunk 21]
```diff
@@ -2914,7 +2871,6 @@ void BlueStore::Blob::split(Collection *coll, uint32_t blob_offset, Blob *r)
     &(r->used_in_blob));
 
   lb.split(blob_offset, rb);
-  dirty_bc().split(get_cache(), blob_offset, r->dirty_bc());
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:2917]: yes!

File: src/os/bluestore/BlueStore.cc [hunk 22]
```diff
@@ -15431,36 +15410,39 @@ void BlueStore::_do_write_small(
 	  }
           logger->inc(l_bluestore_write_small_pre_read);
 
-	  _buffer_cache_write(txc, b, b_off, bl,
-			      wctx->buffered ? 0 : Buffer::FLAG_NOCACHE);
+          bufferlist without_padding;
+          without_padding.substr_of(bl, head_pad, bl.length() - head_pad);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc]: True, I don't seem to understand the existence of padding here. I understand head_read and tail_read as a way of saying "read remainder of block so we don't overwrite with garbage" and I imagine padding is "put zeros where no data should exists". Nevertheless, I moved buffer_cache_write to the start of the `if block` so that we lose complexity of offset calculations`.

File: src/os/bluestore/BlueStore.cc [hunk 23]
```diff
@@ -2878,6 +2878,19 @@ void BlueStore::Blob::split(Collection *coll, uint32_t blob_offset, Blob *r)
 	   << "    and " << *r << dendl;
 }
 
+
+void BlueStore::Blob::maybe_prune_tail() {
+  if (get_blob().is_shared()) {
+    return;
+  }
+
+  if (!get_blob().is_compressed() && get_blob().can_prune_tail()) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:2818]: you might want to move !is_shared() && !is_compressed() check into can_prune_tail() and simplify the code here

File: src/os/bluestore/BlueStore.h (hunk 19 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.h:1650]: Perhaps I'm missing something but what's about the following scenario:
1. Write to Onode1 (aka O1) is being deferred. It belongs to Collection1(aka C1).
2.  O1 is cloned to O2. C1 gets O2 in deferred_seq_dependencies.
3. C1 is split into C1_1 and C1_2. O1 is moved into C1_2 while deferred_seq_dependencies content is left in C1_1.
4. finish_write is called on O1 and it uses deferred_seq_dependencies from O1->c (=C1_2) to notify O2. But it's not there.

File: src/os/bluestore/BlueStore.cc (hunk 20 above)
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:5088]: wait... there should be no "writing" buffers in the cache. But they're still in bc.buffer_map. So there are two issues with this code now:
1. it will wrongly remove the buffer from the source cache. Looking at a.g. LruBufferCacheShard::_rm method implementation I can see no protection from removing a buffer which doesn't belong to it.
2. it will wrongly add it to the destination one. That's evidently an inconsistent state - "writing" buffers shouldn't be there.

File: src/os/bluestore/BlueStore.h (hunk 19 above)
Comment (pereman2) [reply src/os/bluestore/BlueStore.h:1650]: I Deferred IO must be completed before splitting cache as mentioned in this line:
```c++

int BlueStore::_split_collection(TransContext *txc,
				CollectionRef& c,
				CollectionRef& d,
				unsigned bits, int rem)
{
  dout(15) << __func__ << " " << c->cid << " to " << d->cid << " "
	   << " bits " << bits << dendl;
  std::unique_lock l(c->lock);
  std::unique_lock l2(d->lock);
  int r;

  // flush all previous deferred writes on this sequencer.  this is a bit
  // heavyweight, but we need to make sure all deferred writes complete
  // before we split as the new collection's sequencer may need to order
  // this after those writes, and we don't bother with the complexity of
  // moving those TransContexts over to the new osr.
  _osr_drain_preceding(txc);

```

To be fair, I don't like how these semantics are hard to parse mentally.

**Comment by ifed01** [comment]
Hi Pere,
I've just submitted a PR against your branch with BufferSpace rework. It takes writings handling out of BufferSpace into a centralized entry belonging to BlueStore. Making the code a bit more modular and eliminating the need to toss writings/deferred_seq lists when cloning/splitting/etc.
Plus a couple more UTs.
Please take a look and cherry pick to your PR if you find it meaningful.

**Comment by pereman2** [comment]
Pasting here latest results after merging igor's fix for deferred writes:

cpu/mem
image

randrw

image

randwrites

image

**Review by ifed01 — APPROVED** [review]

**Comment by pereman2** [comment]
https://pulpito.ceph.com/pdiazbou-2024-04-29_10:44:48-rados-wip-pereman2-buffer-space-distro-default-smithi
failures look unrelated

**Comment by NitzanMordhai** [comment]
@pereman2 can you please check /a/yuriw-2024-05-21_01:07:55-rados-wip-yuri7-testing-2024-05-20-1227-distro-default-smithi/7717735/teuthology.log

**Comment by pereman2** [comment]
> @pereman2 can you please check /a/yuriw-2024-05-21_01:07:55-rados-wip-yuri7-testing-2024-05-20-1227-distro-default-smithi/7717735/teuthology.log

taking a look, thanks

**Comment by pereman2** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by pereman2** [comment]
https://pulpito.ceph.com/pdiazbou-2024-06-13_14:12:47-rados-wip-pereman2-buffer-space-distro-default-smithi/
